### PR TITLE
JS enhancements for "Add another" further conditions

### PR DESCRIPTION
**Pull request by benswannack**
## Context
We added the ability to add and remove further conditions previously. This required refreshing the page each time though. This PR adds progressive JS enhancements for that feature to avoid page reloads.

## Changes proposed in this pull request
Add Rails UJS
Use remote form to submit using JS

## Guidance to review
Submit does not work well with validations

## Link to Trello card
https://trello.com/c/u0fZBMtg/3613-implement-js-progressive-enhancements-to-infinite-offer-conditions

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/helpers/add_another_button_helper.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module AddAnotherButtonHelper
```
Comment (benswannack) [inline app/helpers/add_another_button_helper.rb:1]: Add some specs for this

File: app/views/provider_interface/offer/conditions/add_another.js.erb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<%= render 'provider_interface/offer/conditions/set_add_button_visibility' %>
```
Comment (benswannack) [inline app/views/provider_interface/offer/conditions/add_another.js.erb:1]: Add a system spec to cover the JS behaviour

File: app/controllers/provider_interface/offer/conditions_controller.rb [hunk 3]
```diff
@@ -61,9 +72,15 @@ def submit_form(action:)
           @wizard.remove_empty_conditions!
           @wizard.save_state!
 
-          redirect_to [action, :provider_interface, @application_choice, :offer, @wizard.next_step]
+          respond_to do |format|
+            format.js { render js: "window.location.replace('#{polymorphic_url([action, :provider_interface, @application_choice, :offer, @wizard.next_step])}')" }
+            format.html { redirect_to [action, :provider_interface, @application_choice, :offer, @wizard.next_step] }
```
Comment (benswannack) [inline app/controllers/provider_interface/offer/conditions_controller.rb:77]: Commonize the redirect URL

File: app/controllers/provider_interface/offer/conditions_controller.rb [hunk 4]
```diff
@@ -61,9 +72,15 @@ def submit_form(action:)
...
+          end
         else
-          render action
+          respond_to do |format|
+            format.js { render js: 'window.location.reload()' }
```
Comment (benswannack) [inline app/controllers/provider_interface/offer/conditions_controller.rb:81]: This doesn't work:
- The HTML method relies on running the validations on the wizard, and then rendering the view with the already validated wizard object
- This just redirects to the new action, which makes a new (unvalidated) wizard.

**Comment by paulrobertlloyd** [comment]
Didn't we add UJS before and then remove it for some reason? Faint memories… @duncanjbrown

**Comment by benswannack** [comment]
> Didn't we add UJS before and then remove it for some reason? Faint memories… @duncanjbrown

When we added it before, it made all forms `remote: true` by default. That was since changed ~fixed~ in rails, so the other forms will still be submitted normally.

**Comment by paulrobertlloyd** [comment]
@benswannack 😅 – I’m not going mad then!

**Comment by benswannack** [comment]
Closing this for now, because there seems to be a simpler way that solves all the problems.

#4721

**Pull request closed without merging**